### PR TITLE
Devoncarew sharing

### PR DESCRIPTION
**Pull request by devoncarew**
Not ready for review

PR to add initial sharing support. This adds a 'Share' button which will create an anonymous copy of the pad on gist.github.com. Also adds an editable header, and a `New...` button.

screen shot 2015-04-03 at 3 42 33 pm

### CODE REVIEW

**Comment by kasperpeulen** [comment]
woooooooooowwwwwwwwww

:D 

+1

**Comment by lukechurch** [comment]
Great @devoncarew let me know when you would like me to take a look

**Comment by devoncarew** [comment]
OK, @kasperpeulen, @lukechurch, ready for review now:

- moved some classes out of playground.dart and into actions.dart, in order to reduce the size of playground
- created a toast class, to display transitory 'ok' and error messages
- wrote code to persist unsaved changes (i.e., the don't lose work issue); that code is disabled for this PR - will be debugged and enabled for a later one
- added a `New...` button, and a `Share` button
- the gist description at the top of the screen is now editable
- added a data binding library as well as tests. It makes it easy to bind sources to targets, and makes for writing less code. So the gist's property `description` is bound to the UI element for displaying and editing the description
- abstracted the gist loading and saving code; you can now decorate both the loaded gist and post-process the gist as it's saving. we use this to make sure the file names are what we'd want, and to extract the html body out of the index.html file
- added a `MuteableGist` abstraction over a `Gist`. It can be edited w/o changing the underlying gist. This is useful to know when a gist is dirty, or to revert all changes to a gist since it was last saved. The MuteableGist also exposes bindable properties, so the text editor's can bind directly to the MuteableGist's fields.
- removed the html and css samples - the starting sample now just has dart code

Comments welcome! Sorry for the size of the PR -

**Comment by kasperpeulen** [comment]
Wow, such an awesome PR :) give me some hours to go through this ^^

File: lib/elements/elements.dart [hunk 1]
```diff
@@ -279,3 +289,65 @@ class DOverlay extends DElement {
...
   }
 }
+
+class DContentEditable extends DElement {
+  DContentEditable(Element element) : super(element) {
+    setAttr('contenteditable', 'true');
+
+    element.onKeyPress.listen((e) {
+      if (e.keyCode == 13) {
+        e.preventDefault();
```
Comment (kasperpeulen) [inline lib/elements/elements.dart]: maybe instead of 13, `KeyCode.ENTER` or `printKeyEvent(e) == "enter"`

File: web/index.css [hunk 2]
```diff
@@ -53,6 +53,7 @@ header {
   color: #ccc;
   font-size: 14pt;
   text-align: center;
```
Comment (kasperpeulen) [inline web/index.css]: `text-align: center` is two times declared

**Comment by kasperpeulen** [comment]
image

I don't really like standard border that contenteditable elements have when they are focused. Maybe we can style this is a bit, or just remove it with:

```
[contenteditable]:focus {
  outline: 0 solid transparent;
}
```

File: lib/elements/elements.css [hunk 3]
```diff
@@ -194,3 +194,21 @@
...
+  bottom: 0;
+  box-shadow: rgba(255, 255, 255, 0.1) 0 1px 0, rgba(0, 0, 0, 0.8) 0 1px 7px 0 inset;
+  min-width: 12em;
+  max-width: 40em;
+  background-color: inherit;
+
+  z-index: 10;
+  transition: 400ms;
+}
+
```
Comment (kasperpeulen) [inline lib/elements/elements.css]: maybe style this white or green to make it more visible ? I didn't notice the toast message, only after I knew it should exist and I was looking for it, I saw it.

File: web/index.html [hunk 4]
```diff
@@ -94,7 +102,7 @@
           <div class="inner"></div>
         </div>
         <div class="view" flex two layout vertical>
-          <div class="header" layout horizontal wrap>
+          <div class="header small" layout horizontal wrap>
             <li><a id="consoletab" selected><span>Console</span></a></li>
```
Comment (kasperpeulen) [inline web/index.html]: not so sure about this, seems a bit odd that the tab text (Console, Documentation) is smaller than the text in the documentation

File: lib/elements/elements.dart (hunk 1 above)
Comment (devoncarew) [inline lib/elements/elements.dart]: changed to `KeyCode.ENTER`

File: web/index.css (hunk 2 above)
Comment (devoncarew) [inline web/index.css]: oops :)

**Comment by devoncarew** [comment]
> I don't really like standard border that contenteditable elements have when they are focused. Maybe we can style this is a bit, or just remove it with:

Done - it looks better.

File: lib/elements/elements.css (hunk 3 above)
Comment (devoncarew) [inline lib/elements/elements.css]: I lightened it up a bit, and moved it to the upper right to be more visible.

File: web/index.html (hunk 4 above)
Comment (devoncarew) [inline web/index.html]: I'm trying to make that tab area appear secondary in comparison to the main dart/html/css one. I'll dial the font size up slightly.

**Comment by devoncarew** [comment]
screen shot 2015-04-06 at 5 53 39 pm
screen shot 2015-04-06 at 5 53 48 pm
screen shot 2015-04-06 at 5 54 00 pm

File: lib/actions.dart [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library dartpad.actions;
+
+import 'dart:html';
+
+import 'dart_pad.dart';
+import 'elements/elements.dart';
+//import 'sharing/gist_storage.dart';
```
Comment (lukechurch) [inline lib/actions.dart]: snip

File: lib/actions.dart [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    _button.onClick.listen((e) => _handleButtonPress());
+  }
+
+  void _handleButtonPress() {
+    if (_gist.dirty) {
+      if (!window.confirm('Discard changes to the current pad?')) return;
+    }
+
+    //_gistStorage.clearStoredGist();
+
```
Comment (lukechurch) [inline lib/actions.dart]: Cold you put in a bit more about your intentions with the alternative mode that is commented out?

File: lib/actions.dart [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+  void _handleButtonPress() {
+    if (ga != null) ga.sendEvent('main', 'share');
+
+    gistLoader.createAnon(_gist.createGist()).then((Gist newGist) {
+      _gistContainer.overrideNextRoute(newGist);
+      router.go('gist', {'gist': newGist.id});
+      DToast.showMessage('Created ${newGist.id}');
+    }).catchError((e) {
+      String message = 'Error saving gist: ${e}';
+      DToast.showMessage(message);
```
Comment (lukechurch) [inline lib/actions.dart]: Send an analytics 'task failure' ping? I'd quite like us to get into the habit of measuring how often things are failing.

File: lib/playground.dart [hunk 8]
```diff
@@ -80,11 +92,29 @@ class Playground {
...
+    bind(editableGist.property('description'), titleEditable.textProperty);
+
+    // Update the ID on changes.
+    AnchorElement idAnchor = querySelector('header .header-gist-id a');
+    bind(editableGist.property('id'), (val) => idAnchor.text = val);
+    bind(editableGist.property('html_url'), (val) {
+      idAnchor.href = val == null ? '' : val;
+    });
+
+    Throttler throttle = new Throttler(const Duration(milliseconds: 100));
```
Comment (lukechurch) [inline lib/playground.dart]: Comment out the object to correspond to the usage?

Could you add a short comment at the top of the disabled block?

File: lib/playground.dart [hunk 9]
```diff
@@ -95,31 +125,25 @@ class Playground {
...
-      }
-    }
-
-    _setGistDescription(null);
-    _setGistId(null, null);
-
-    context.dartSource = sample.dartCode;
-    context.htmlSource = sample.htmlCode;
-    context.cssSource = sample.cssCode;
+//    String path = window.location.pathname;
```
Comment (lukechurch) [inline lib/playground.dart]: As above

**Comment by kasperpeulen** [comment]
@devoncarew looks and works perfect now imo, the animation of the toast looks also more fluent now

I haven't gone through all the technical details of the gist, will look at that tomorrow.

**Comment by lukechurch** [comment]
@devoncarew any reason to not push this to a named GAE version? It would make reviewing the HTML/CSS a bit easier to have something to play with. If you're ok, I'll push a special version of it.

**Comment by devoncarew** [comment]
@lukechurch, feel free to push to a named version

File: lib/actions.dart (hunk 5 above)
Comment (devoncarew) [inline lib/actions.dart]: done

File: lib/actions.dart (hunk 7 above)
Comment (devoncarew) [inline lib/actions.dart]: done; sending it as a GA exception type

File: lib/playground.dart (hunk 8 above)
Comment (devoncarew) [inline lib/playground.dart]: done

File: lib/playground.dart (hunk 9 above)
Comment (devoncarew) [inline lib/playground.dart]: done

**Comment by devoncarew** [comment]
> Could you put in a bit more about your intentions with the alternative mode that is commented out?

Current idea is:

- as a gist is edited (after a brief ~300ms reconciliation time), a copy of the gist is stored in the browser's `localStorage`. That's the gist's ID and the current state of it's files. Only a single gist is stored; saving a gist overwrites any previous gist.
- when we get a request to load a gist (the user browses to a page; opens dartpad in a new tab; accidently hits back in the browser and forward again to recover), instead of loading the gist from github, we first check if a gist with an identical id is available in localstorage. If so, we load from localstorage instead.

**Comment by devoncarew** [comment]
Available to dogfood here: http://sharing.dart-pad.appspot.com.

**Comment by lukechurch** [comment]
Thanks @devoncarew for the dogfood - I'm not entirely convinced by the UI for the toast. It's very pretty but looks to me like it's embedded in the page as an editable region - it doesn't look like it's above the page which a toast should do.

**Comment by kasperpeulen** [comment]
About the html page saved to the gist:
```
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>${gist.description}</title>
  </head>

  <body>
    ${htmlFile.content}
  </body>
</html>
```

Maybe also add the links to the stylesheet and the dartfile, like ?

```
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>${gist.description}</title>
    <link rel="stylesheet" href="styles.css">
  </head>

  <body>
    ${htmlFile.content}
  <script type="application/dart" src="main.dart"></script>
  </body>
</html>
```

**Comment by devoncarew** [comment]
@kasperpeulen, +1 to links to the dart and css! I'll add.

@lukechurch, also agree re: the toast comments. I'll play with the UI a bit to try and separate the toast from the page.

**Comment by lukechurch** [comment]
lgtm! As discussed lets bring this in and iterate, but lets make sure https://github.com/dart-lang/dart-pad/issues/285 gets addressed before our next push to prod.

Really happy to see this

**Comment by kasperpeulen** [comment]
lgtm as well, can't wait to see this in prod :)

**Pull request merged**